File: 2016/9/solution.js
const fs = require("fs");
const path = require("path");

const input = fs.readFileSync(
  path.join(__dirname, "input.txt"),
  "utf-8",
  (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  }
);

const decompress = (code) => {
  let i = 0;
  let res = 0;
  while (i < code.length) {
    while (i < code.length && code[i] !== "(") {
      res++;
      i++;
    }
    if (i >= code.length) {
      break;
    }
    let j = i;
    while (code[j] !== ")") {
      j++;
    }
    const marker = code.slice(i + 1, j);
    const nums = marker.split("x").map(Number);
    res += nums[0] * nums[1];
    i = j + 1 + nums[0];
  }
  return res;
};

const decompress2 = (code) => {
  let i = 0;
  let res = 0;
  while (i < code.length) {
    while (i < code.length && code[i] !== "(") {
      res++;
      i++;
    }
    if (i >= code.length) {
      break;
    }
    let j = i;
    let balance = 1;
    while (code[j] !== ")") {
      j++;
    }
    const marker = code.slice(i + 1, j);
    const nums = marker.split("x").map(Number);
    res += decompress2(code.slice(j + 1, j + 1 + nums[0])) * nums[1];
    i = j + 1 + nums[0];
  }
  return res;
};

console.log(decompress(input));
console.log(decompress2(input));

File: 2016/9/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := string(data)
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseMarker(s string) (int, int) {
	parts := strings.Split(s, "x")
	n, err := strconv.Atoi(parts[0])
	if err != nil {
		panic(err)
	}
	k, err := strconv.Atoi(parts[1])
	if err != nil {
		panic(err)
	}
	return n, k
}

func decompress(s string, recursive bool) int {
	res := 0
	i := 0
	for i < len(s) {
		if s[i] != '(' {
			res++
		} else {
			j := i + 1
			for s[j] != ')' {
				j++
			}
			n, k := parseMarker(s[i+1 : j])
			if recursive {
				res += decompress(s[j+1:j+n+1], recursive) * k
			} else {
				res += n * k
			}
			i = j + n
		}
		i++
	}
	return res
}

func part1(s string) int {
	return decompress(s, false)
}

func part2(s string) int {
	return decompress(s, true)
}
